src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1849,39 +1849,94 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
             ResultType = InferResultNullability(binary, leftType, rightType);
 
             BinaryOperatorKind op = binary.OperatorKind.Operator();
+
+            // learn from non-null constant
+            BoundExpression operandComparedToNonNull = null;
+            if (skipImplicitNullableConversions(binary.Left).ConstantValue?.IsNull == false)
+            {
+                operandComparedToNonNull = binary.Right;
+            }
+            else if (skipImplicitNullableConversions(binary.Right).ConstantValue?.IsNull == false)
+            {
+                operandComparedToNonNull = binary.Left;
+            }
+
+            if (operandComparedToNonNull != null)
```

Comment:
What happens in the case `if (null == null)`?

Reply:
Nothing will happen because there is no slot to update for either `null`.
There is however a scenario that is a bit weird:
```C#
const string? nullable = null;
const string? nonNullable = "";
if (nullable == nonNullable)
```

We could say that we don't update constants, or maybe we need to distinguish `null` from other null-valued constants...